ams/server/src/main/java/com/netease/arctic/server/table/TagConfiguration.java

```diff
@@ -65,14 +85,26 @@ public String propertyName() {
       return propertyName;
     }
 
+    protected abstract Duration periodDuration();
+
     /**
-     * Obtain the trigger time for creating a tag, which is the idea time of the last tag before the
-     * check time.
+     * Obtain the trigger time for creating a tag, which is the ideal time of the last tag before
+     * the check time.
      *
      * <p>For example, when creating a daily tag, the check time is 2022-08-08 11:00:00 and the
      * offset is set to be 5 min, the idea trigger time is 2022-08-08 00:05:00.
+     *
+     * <p>For example, when creating a daily tag, the offset is set to be 30 min, if the check time
+     * is 2022-08-08 02:00:00, the ideal trigger time is 2022-08-08 00:30:00; if the check time is
+     * 2022-08-09 00:20:00 (before 00:30 of the next day), the ideal trigger time is still
+     * 2022-08-08 00:30:00.
      */
-    public abstract long getTagTriggerTime(LocalDateTime checkTime, int triggerOffsetMinutes);
+    public abstract LocalDateTime getTagTriggerTime(
+        LocalDateTime checkTime, int triggerOffsetMinutes);
+
+    public LocalDateTime normalizeToTagTime(LocalDateTime triggerTime, int triggerOffsetMinutes) {
```

Comment:
We had better add comments to this public method so that other developers can better understand it.